matrix/plugins/glitch/actions.py

```diff
@@ -0,0 +1,69 @@
+import asyncio
+import logging
+from typing import Any
+from functools import wraps
+
+import enforce
+from juju.model import Model
+from juju.unit import Unit
+
+from matrix.utils import Singleton
+
+log = logging.getLogger("glitch")
+
+
+class _Actions(dict, metaclass=Singleton):
+    """
+    Interface for registering actions, translating text strings into
+    action functions, and performing automagic around our action
+    functions.
+
+    """
+    def decorate(self, func):
+        """
+        Register an action. Return a function that accepts a set of objects,
+        and iterates over that set, running the registered action on each
+        object in that set.
+
+        """
+        @enforce.runtime_validation
+        @wraps(func)
+        async def wrapped(model, objects, **kwargs):
+            for obj in objects:
+                await func(model, obj, **kwargs)
```

Comment:
I think it would be better to have the actions explicitly accept sets of objects, rather than implicitly converting them by calling the action multiple times, because some implementations (e.g., `juju run` or `juju run-action`) can handle multiple objects natively, which would be more efficient.

However, since the calls will be dispatched async, it's not a big enough deal to consider a blocker.